ui/app/src/components/shared/CreateUpdateResource/CreateUpdateResource.tsx

```diff
@@ -0,0 +1,143 @@
+import { NotificationsContext } from '../notifications/NotificationsContext';
+import { ResourceForm } from './ResourceForm';
+import { SelectTemplate } from './SelectTemplate';
+
+interface CreateUpdateResourceProps {
+  isOpen: boolean,
+  onClose: () => void,
+  resourceType: ResourceType,
+  parentResource?: Workspace | WorkspaceService
+}
+
+interface PageTitle {
+  selectTemplate: string,
+  resourceForm: string,
+  creating: string
+}
+
+const creatingIconClass = mergeStyles({
+  fontSize: 100,
+  height: 100,
+  width: 100,
+  margin: '0 25px',
+  color: 'deepskyblue',
+  padding: 20
+});
+
+export const CreateUpdateResource: React.FunctionComponent<CreateUpdateResourceProps> = (props: CreateUpdateResourceProps) => {
+  const [page, setPage] = useState('selectTemplate' as keyof PageTitle);
+  const [selectedTemplate, setTemplate] = useState('');
+  const [deployOperation, setDeployOperation] = useState({} as Operation);
+  const opsContext = useContext(NotificationsContext);
+
+  // Render a panel title depending on sub-page
+  const pageTitles: PageTitle = {
+    selectTemplate: 'Choose a template',
+    resourceForm: 'Create a new ' + props.resourceType,
+    creating: ''
+  }
+
+  // Construct API path for templates of specified resourceType
+  let templatesPath;
+  switch (props.resourceType) {
+    case ResourceType.Workspace:
+      templatesPath = ApiEndpoint.WorkspaceTemplates; break;
+    case ResourceType.WorkspaceService:
+      templatesPath = ApiEndpoint.WorkspaceServiceTemplates; break;
+    case ResourceType.SharedService:
+      templatesPath = ApiEndpoint.SharedServiceTemplates; break;
+    case ResourceType.UserResource:
+      if (props.parentResource) {
+        templatesPath = `${ApiEndpoint.WorkspaceServiceTemplates}/${props.parentResource.templateName}/${ApiEndpoint.UserResourceTemplates}`; break;
+      } else {
+        throw Error('Parent workspace service must be passed as prop when creating user resource.');
+      }
+    default:
+      throw Error('Unsupported resource type.');
+  }
+
+  // Construct API path for resource creation
+  let resourcePath;
+  switch (props.resourceType) {
+    case ResourceType.Workspace:
+      resourcePath = ApiEndpoint.Workspaces; break;
+    case ResourceType.SharedService:
+      resourcePath = ApiEndpoint.SharedServices; break;
+    default:
+      if (!props.parentResource) {
+        throw Error('A parentResource must be passed as prop if creating a workspace-service or user-resource');
+      }
+      resourcePath = `${props.parentResource.resourcePath}/${props.resourceType}`;
+  }
+
+  const selectTemplate = (templateName: string) => {
+    setTemplate(templateName);
+    setPage('resourceForm');
+  }
+
+  const resourceCreating = (operation: Operation) => {
+    setDeployOperation(operation);
+    setPage('creating');
+    // Add deployment operation to notifications operation poller
+    opsContext.addOperation(operation);
+  }
+
+  const navigate = useNavigate();
```

Comment:
nit but i think hooks should be defined just inside the function signature

Reply:
You're right!